File: backend/app/models/feedback.py
from pydantic import BaseModel, Field

from app.models.rubric import RubricCriterion


class IngestInput(BaseModel):
    file_path: str
    file_type: str


class IngestOutput(BaseModel):
    raw_text: str
    word_count: int
    paragraph_count: int
    detected_sections: list[str]
    language: str = "en"


class GrammarError(BaseModel):
    type: str
    message: str
    offset: int


class ParagraphCoherence(BaseModel):
    paragraph_index: int
    score: int = Field(ge=1, le=5)
    issues: list[str]
    suggestions: list[str]


class NLPAnalysisOutput(BaseModel):
    grammar_errors: list[GrammarError]
    grammar_error_count: int
    avg_sentence_length: float
    passive_voice_ratio: float
    ttr: float
    paragraph_coherence: list[ParagraphCoherence]
    overall_flow_score: float
    flow_comments: str


class CriterionScore(BaseModel):
    criterion: str
    max_points: float
    awarded_points: float
    justification: str
    strengths: list[str]
    improvements: list[str]
    evidence_quotes: list[str] = Field(default_factory=list)


class RubricMatchOutput(BaseModel):
    rubric_scores: list[CriterionScore]
    total_score: float
    total_possible: float


class FeedbackReport(BaseModel):
    overall_summary: str
    total_score: float
    total_possible: float
    grade_breakdown: list[CriterionScore]
    top_strengths: list[str]
    priority_improvements: list[str]
    suggested_next_steps: list[str]


class FeedbackRunRequest(BaseModel):
    file_id: str
    rubric_id: str


class UploadResponse(BaseModel):
    file_id: str
    filename: str
    file_path: str
    file_type: str
    size_bytes: int


class FeedbackRecord(BaseModel):
    feedback_id: str
    rubric_id: str
    file_id: str
    report: FeedbackReport
    ingest: IngestOutput
    analysis: NLPAnalysisOutput
    rubric_match: RubricMatchOutput


class ErrorEnvelope(BaseModel):
    error: str
    detail: str | None = None

File: frontend/my-react-app/src/types/feedback.ts
export interface RubricCriterion {
  name: string
  max_points: number
  description: string
}

export interface CriterionScore {
  criterion: string
  max_points: number
  awarded_points: number
  justification: string
  strengths: string[]
  improvements: string[]
  evidence_quotes?: string[]
}

export interface FeedbackReport {
  overall_summary: string
  total_score: number
  total_possible: number
  grade_breakdown: CriterionScore[]
  top_strengths: string[]
  priority_improvements: string[]
  suggested_next_steps: string[]
}

export interface RubricResponse {
  rubric_id: string
  criteria: RubricCriterion[]
}

export interface UploadResponse {
  file_id: string
  filename: string
  file_path: string
  file_type: string
  size_bytes: number
}

export interface MCPToolExecution {
  tool_name: string
  payload: Record<string, unknown>
  result: Record<string, unknown>
}

export interface MCPBridgeRequest {
  user_prompt: string
  session_id?: string
  file_id?: string
  rubric_id?: string
  raw_text?: string
  rubric?: RubricCriterion[]
  max_steps?: number
}

export interface MCPBridgeResponse {
  session_id: string
  assistant_message: string
  intent: 'analyze_text' | 'score_rubric' | 'full_feedback'
  planned_tools: string[]
  executed_steps: MCPToolExecution[]
  artifacts: Record<string, unknown>
}
